src/chains/ethereum/ethereum/src/api.ts

```diff
@@ -2877,6 +2877,184 @@ export default class EthereumApi implements Api {
+  ): Promise<Ethereum.FeeHistory> {
+    const blockchain = this.#blockchain;
+    const PAD_PRECISION = 16;
+    const PRECISION_FLOAT = 1e14;
+    const PRECISION_BIG_INT = BigInt(1e16);
+
+    // if newestBlock is a tag, we already keep a ref in the Block Manager.
+    let newestBlockNumber;
+    const newBlock = await blockchain.blocks.getBlockByTag(newestBlock as Tag);
+
+    // If newestBlock is not a tag we can just use the newestBlock number.
+    if (newBlock) {
+      newestBlockNumber = newBlock.header.number.toNumber();
+    } else {
+      newestBlockNumber = Quantity.from(newestBlock).toNumber();
+    }
+
+    // blockCount > newestBlock is technically valid but we cannot go past the Genesis Block.
+    const totalBlocks = Math.min(
+      Quantity.toNumber(blockCount),
+      newestBlockNumber + 1
+    );
+
+    // Cut out early if no range is given.
+    if (totalBlocks === 0) {
+      return {
+        oldestBlock: Quantity.from(newestBlock).toString(),
+        baseFeePerGas: undefined,
+        gasUsedRatio: null,
+        reward: undefined
+      };
+    }
+
+    // blockCount is inclusive of newestBlock
+    const oldestBlockNumber = newestBlockNumber - (totalBlocks - 1);
+
+    const baseFeePerGas = new Array(totalBlocks);
+    const gasUsedRatio = new Array(totalBlocks);
+    let reward;
+    if (rewardPercentiles.length > 0) {
+      reward = new Array(totalBlocks);
+    }
+
+    let currentBlockNumber = oldestBlockNumber;
+    let currentBlock;
+
+    while (currentBlockNumber <= newestBlockNumber) {
+      currentBlock = await blockchain.blocks.get(
+        Quantity.toBuffer(currentBlockNumber)
+      );
+
+      const currentPosition = currentBlockNumber - oldestBlockNumber;
+
+      baseFeePerGas[currentPosition] =
+        currentBlock.header.baseFeePerGas || Quantity.Zero;
+
+      gasUsedRatio[currentPosition] = Number(
+        `0.${(
+          (currentBlock.header.gasUsed.toBigInt() * PRECISION_BIG_INT) /
+          currentBlock.header.gasLimit.toBigInt()
+        )
+          .toString()
+          .padStart(PAD_PRECISION, "0")}`
+      );
+
+      if (reward) {
+        const transactions = currentBlock.getTransactions();
+
+        // If there are no transactions, all reward percentiles are 0.
+        if (transactions.length === 0) {
+          reward[currentPosition] = rewardPercentiles.map(() => {
+            return Quantity.Zero;
+          });
+        } else {
+          // For all transactions, effectiveGasReward = normalized fee per unit of gas
+          // earned by the miner regardless of transaction type
+          const effectiveRewardAndGasUsed = (
+            await Promise.all(
+              transactions.map(async tx => {
+                const receipt = (
+                  await blockchain.transactionReceipts.get(tx.hash.toBuffer())
+                ).toJSON(currentBlock, tx, blockchain.common);
+
+                const baseFeePerGas = currentBlock.header.baseFeePerGas
```

Comment:
This value should be calculated outside of the loop.

Reply:
:+1: